### PR TITLE
Source flags and src scope

### PR DESCRIPTION
**Pull request by Naktakala**
This PR changes quite a bit of files. Please read the highlights below before reviewing to better comprehend the impact of the minor items.

# Highlights
- `SweepChunk*` changed to `std::shared_ptr<SweepChunk>` in `SetSweepChunk` routine, and references are now passed where applicable, instead of the raw pointer.
- `SourceFlags` is now an enum.
- Some renaming to keep consistency with nodes-vs-DOFs
- Ad-hoc renamings (not a lot)
- Somewhat of a cleanup of the source routines (both `SetSource` and `SetKSource`.
- Reorganization of GMRES
- `KSPDataContext` and `LBSMatrixAction_Ax` has now been appended to the `LinearBoltzmann` namespace
- Replaced all the pointers in `KSPDataContext` with references and added a constructor
- Removed an old unused `KSPMonitorNPT` routine

# Details regarding source flags
- Adding the material-source, the scattering-source and the fission-source to the `q_moments`-vector is an ability included in the `SetSource` routine. The first vexing thing was that we had this strange yes-no logic for only two of these items where we had an `apply_mat_src`-flag and a `suppress_phi_old`-flag, so if you wanted both the material-source and the scattering-source you had to say `SetSource(groupset, true, false)`. The second thing was that fission was operated via the `suppress_phi_old`-flag. These flags are now all consolidated into an integer bit-set which will support further future options.

### Old code example GMRES
```
//Computing b
SetSource(groupset, SourceFlags::USE_MATERIAL_SOURCE, SourceFlags::SUPPRESS_PHI_OLD);

//Inside matrix action
SetSource(groupset, LinearBoltzmann::SourceFlags::USE_DLINV_SOURCE, false);
```
### New code example GMRES
```
//Computing b
SourceFlags rhs_src_scope = APPLY_MATERIAL_SRC;
SetSource(groupset, rhs_src_scope);

//Inside matrix action
SourceFlags lhs_src_scope = APPLY_SCATTER_SOURCE | APPLY_FISSION_SOURCE;
SetSource(groupset, lhs_src_scope);
```

GMRES also now gets called with the flags for `lhs_src_scope` and `rhs_src_scope` adding a lot of versatility to any call to GMRES. Examples:
- One can do GMRES as usual `rhs_src_scope = APPLY_MATERIAL_SRC` and `lhs_src_scope = APPLY_SCATTER_SOURCE | APPLY_FISSION_SOURCE`
- One can do GMRES to purely converge cyclic angular fluxes `rhs_src_scope = APPLY_MATERIAL_SRC | APPLY_SCATTER_SOURCE | APPLY_FISSION_SOURCE` and `lhs_src_scope = NO_FLAGS_SET`

One can also theoretically call `LBSMatrixAction_Ax`-standalone although this is not ready yet since it get the context from a matrix shell. But this will soon be split into an operator-action section and the matrix action.

### CODE REVIEW

File: ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc [hunk 1]
```diff
@@ -156,8 +158,8 @@ void KEigenvalue::Solver::PowerIteration(LBSGroupset& groupset)
     chi_log.ProcessEvent(source_event_tag,
                          ChiLog::EventOperation::AVERAGE_DURATION);
   size_t num_angles = groupset.quadrature->abscissae.size();
-  long int num_unknowns = (long int)glob_dof_count*
-                          (long int)num_angles*
+  long int num_unknowns = (long int)glob_node_count *
```
Comment (ragusa) [inline ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc]: is this an example where C++ casting should be used?

File: ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc [hunk 2]
```diff
@@ -121,14 +115,14 @@ void KEigenvalue::Solver::SetKSource(LBSGroupset& groupset,
           q_mom[g] += fission_g;
 
           // ----- Contribute precursors
-          precursor_g = 0.0;
+          double precursor_g = 0.0;
           if ((ell == 0) and (options.use_precursors))
           {
             if ((apply_mat_src) and (xs->J > 0))
```
Comment (ragusa) [inline ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc]: I guess J is the # of precursors. Could we use a better name for that variable than just J?

File: ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc [hunk 3]
```diff
@@ -80,36 +74,36 @@ void KEigenvalue::Solver::SetKSource(LBSGroupset& groupset,
...
+        double* phi_prevp = &phi_prev_local[ir];
 
         // ----- Loop over groupset groups
         for (int g=gs_i; g<=gs_f; g++)
         {
           // ----- Contribute scattering
-          inscat_g = 0.0;
-          if ((ell < xs->transfer_matrix.size()) && (!suppress_phi_old) )
+          double inscat_g = 0.0;
+          if ((ell < xs->transfer_matrix.size()) && (apply_scatter_src) )
```
Comment (ragusa) [inline ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc]: Why is apply_scatter_src for?

Comment (ragusa) [inline ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc]: spacing and parentheses is not consistent.

File: ChiModules/LinearBoltzmannSolver/IterativeOperations/lbs_setsource.cc [hunk 4]
```diff
@@ -65,54 +68,49 @@ void LinearBoltzmann::Solver::SetSource(LBSGroupset& groupset,
...
-                sigma_sm  = xs->transfer_matrix[ell].rowI_values[g][t];
+                double sigma_sm = xs->transfer_matrix[ell].rowI_values[g][t];
                 inscat_g += sigma_sm * phi_oldp[gprime];
               }
             }
           }//if moment avail
 
           //====================== Apply within-groupset scattering
-          if ((ell < xs->transfer_matrix.size()) && (!suppress_phi_old) )
+          if ((ell < xs->transfer_matrix.size()) && (apply_scatter_src) )
```
Comment (ragusa) [inline ChiModules/LinearBoltzmannSolver/IterativeOperations/lbs_setsource.cc:105]: parenthesis

File: ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc (hunk 2 above)
Comment (Naktakala) [reply ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc]: I agree. This however would require a number of changes not in the scope of this PR. Would you be ok with me creating an issue for it to the code author (Zach Hardy)?

File: ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc (hunk 3 above)
Comment (Naktakala) [reply ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc]: You probably mean "What is apply_scatter_src for?". It is a means to turn off certain portions of the `SetKSource` routine similar to what the `SetSource` routine does. The usefulness is not immediately apparent since we don't have any other methods other than `PowerIteration` but unless you have a strong objection against I would rather leave it untouched until we either can confirm it is not needed or find a better way to use the routine.

File: ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc (hunk 1 above)
Comment (Naktakala) [reply ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc]: Yes. This code was changed as part of a refactor but I went ahead and changed the obvious c-style casts in the file. Also cleaned it up a bit on some places where clang-tidy complained.

File: ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc (hunk 3 above)
Comment (Naktakala) [reply ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc]: Corrected.

File: ChiModules/LinearBoltzmannSolver/IterativeOperations/lbs_setsource.cc (hunk 4 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/IterativeOperations/lbs_setsource.cc:105]: Corrected

**Review by dominiccaron — APPROVED** [review]

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc [hunk 5]
```diff
@@ -40,65 +44,53 @@ bool LinearBoltzmann::Solver::GMRES(LBSGroupset& groupset,
...
+  //================================================== Get groupset dof sizes
+  size_t groupset_numgrps = groupset.groups.size();
+  auto num_delayed_ang_unknowns = groupset.angle_agg.GetNumberOfAngularUnknowns();
+  size_t local_size = local_node_count * num_moments * groupset_numgrps +
+                      num_delayed_ang_unknowns.first;
+  size_t globl_size = glob_node_count * num_moments * groupset_numgrps +
+                      num_delayed_ang_unknowns.second;
+
+  //================================================== Create PETSc vectors
+  phi_new = chi_math::PETScUtils::CreateVector(static_cast<int64_t>(local_size),
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc:56]: Why are we casting these values to int64?

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc [hunk 6]
```diff
@@ -40,65 +44,53 @@ bool LinearBoltzmann::Solver::GMRES(LBSGroupset& groupset,
...
-  auto num_ang_unknowns = groupset.angle_agg.GetNumberOfAngularUnknowns();
-  int local_size = local_dof_count*num_moments*groupset_numgrps +
-                   num_ang_unknowns.first;
-  int globl_size = glob_dof_count*num_moments*groupset_numgrps +
-                   num_ang_unknowns.second;
-  MatCreateShell(PETSC_COMM_WORLD,local_size,
-                                  local_size,
-                                  globl_size,
-                                  globl_size,
+  MatCreateShell(PETSC_COMM_WORLD,static_cast<int64_t>(local_size),
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc:72]: As above, why is this cast necessary?

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc [hunk 7]
```diff
@@ -185,8 +180,8 @@ bool LinearBoltzmann::Solver::GMRES(LBSGroupset& groupset,
       chi_log.ProcessEvent(source_event_tag,
                            ChiLog::EventOperation::AVERAGE_DURATION);
     size_t num_angles = groupset.quadrature->abscissae.size();
-    long int num_unknowns = (long int)glob_dof_count*
-                            (long int)num_angles*
+    long int num_unknowns = (long int)glob_node_count *
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc]: Since I believe these are all size_t, why the casts?

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc [hunk 8]
```diff
@@ -204,7 +199,8 @@ bool LinearBoltzmann::Solver::GMRES(LBSGroupset& groupset,
         << chunk_overhead_ratio;
       chi_log.Log(LOG_0)
         << "        Sweep Time/Unknown (ns):       "
-        << sweep_time*1.0e9*chi_mpi.process_count/num_unknowns;
+        << sweep_time*1.0e9*chi_mpi.process_count/
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc:202]: Just a note that at some point, we really need to convert to chrono and duration_cast to avoid things like a random multiply by 1e9.

File: ChiModules/LinearBoltzmannSolver/Tools/ksp_data_context.h [hunk 9]
```diff
@@ -1,16 +1,38 @@
...
-  int last_iteration = -1;
-};
\ No newline at end of file
+  LinearBoltzmann::Solver& solver;
+  SweepChunk&              sweep_chunk;
+  LBSGroupset&             groupset;
+  Vec&                     operating_vector;
+  chi_mesh::sweep_management::SweepScheduler& sweepScheduler;
+  SourceFlags    lhs_scope;
+  size_t last_iteration = -1;
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/Tools/ksp_data_context.h]: Not sure if this is intentional or not. While wraparound is well defined by the standard, I argue that it's bad practice.  If it should accept signed values, use a signed type.

File: ChiModules/LinearBoltzmannSolver/Tools/kspmonitor_npt.cc [hunk 10]
```diff
@@ -81,28 +62,28 @@ KSPConvergenceTestNPT(KSP ksp, PetscInt n, PetscReal rnorm,
...
 
-  if (context->groupset->iterative_method == NPT_GMRES)
+  if (context->groupset.iterative_method == NPT_GMRES)
   {
     if (context->last_iteration == n)
     {
-      if (context->solver->options.write_restart_data)
+      if (context->solver.options.write_restart_data)
       {
         if ((chi_program_timer.GetTime()/60000.0) >
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/Tools/kspmonitor_npt.cc]: Although not part of the pr, we should document the choice of 60000.0.

**Comment by ragusa** [comment]
Good point. I wonder too.

—jean ragusa

multiphysics.engr.tamu.edu
nustem.engr.tamu.edu

On May 21, 2021, at 1:27 PM, wdhawkins ***@***.***> wrote:

﻿

@wdhawkins commented on this pull request.

________________________________

In ChiModules/LinearBoltzmannSolver/Tools/kspmonitor_npt.cc<https://urldefense.com/v3/__https://github.com/chi-tech/chi-tech/pull/131*discussion_r637121088__;Iw!!KwNVnqRv!VGkK47o9epY4KaQsgRN6DM8nXMDRB7PzErjRBqJCw18ftQjGdkZfIRlEr2ZchCNxWw$>:

>    {
     if (context->last_iteration == n)
     {
-      if (context->solver->options.write_restart_data)
+      if (context->solver.options.write_restart_data)
       {
         if ((chi_program_timer.GetTime()/60000.0) >


Although not part of the pr, we should document the choice of 60000.0.

File: ChiModules/LinearBoltzmannSolver/lbs_01d_initparrays.cc [hunk 11]
```diff
@@ -55,14 +55,14 @@ void LinearBoltzmann::Solver::InitializeParrays()
...
-  glob_dof_count = discretization->GetNumGlobalDOFs(flux_moments_uk_man)/GxM;
+  auto& per_node = ChiMath::UNITARY_UNKNOWN_MANAGER;
+  local_node_count = discretization->GetNumLocalDOFs(per_node);
+  glob_node_count = discretization->GetNumGlobalDOFs(per_node);
 
   //================================================== Compute num of unknowns
   int num_grps = groups.size();
   int M = num_moments;
-  unsigned long long local_unknown_count = local_dof_count * num_grps * M;
+  unsigned long long local_unknown_count = local_node_count * num_grps * M;
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/lbs_01d_initparrays.cc:65]: I think this has the potential to cause problems if num_grps * M exceeds int_max.

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc (hunk 5 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc:56]: Because that is what PETSc takes. On that note, we could create an issue to modify the `PETScUtils` so that those utility functions use `PetscInt` instead of `int64_t` (which would be more consistent). What do you think?

File: ChiModules/LinearBoltzmannSolver/lbs_03d_wgdsa.cc [hunk 12]
```diff
@@ -110,7 +110,7 @@ void LinearBoltzmann::Solver::AssembleWGDSADeltaPhiVector(LBSGroupset& groupset,
   int gss = groupset.groups.size();
 
   delta_phi_local.clear();
-  delta_phi_local.resize(local_dof_count*gss,0.0);
+  delta_phi_local.resize(local_node_count * gss, 0.0);
```
Comment (wdhawkins) [inline ChiModules/LinearBoltzmannSolver/lbs_03d_wgdsa.cc:113]: Just an interesting observation that resize followed by fill might be faster than clear and resize.

**Review by wdhawkins — CHANGES_REQUESTED** [review]
My only major concern is with the wraparound in ksp_data_context.

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc (hunk 7 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc]: Fixed.

File: ChiModules/LinearBoltzmannSolver/Tools/ksp_data_context.h (hunk 9 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/Tools/ksp_data_context.h]: Great catch Thx!

File: ChiModules/LinearBoltzmannSolver/lbs_01d_initparrays.cc (hunk 11 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/lbs_01d_initparrays.cc:65]: How to fix?

File: ChiModules/LinearBoltzmannSolver/lbs_03d_wgdsa.cc (hunk 12 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/lbs_03d_wgdsa.cc:113]: What would you like to see here?

File: ChiModules/LinearBoltzmannSolver/Tools/kspmonitor_npt.cc (hunk 10 above)
Comment (Naktakala) [reply ChiModules/LinearBoltzmannSolver/Tools/kspmonitor_npt.cc]: More descriptive constant defined. This was intended to be an input parameter. Will create issue. For now, this is limited to an interval of 60s (60k milliseconds).

File: ChiModules/LinearBoltzmannSolver/lbs_03d_wgdsa.cc (hunk 12 above)
Comment (wdhawkins) [reply ChiModules/LinearBoltzmannSolver/lbs_03d_wgdsa.cc:113]: No fix necessary.  Just an observation.

File: ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc (hunk 5 above)
Comment (wdhawkins) [reply ChiModules/LinearBoltzmannSolver/IterativeMethods/lbs_GMRES.cc:56]: Yes, I think that's probably the correct thing to do.  It sounds like it's at least the consistent thing to do.

File: ChiModules/LinearBoltzmannSolver/lbs_01d_initparrays.cc (hunk 11 above)
Comment (wdhawkins) [reply ChiModules/LinearBoltzmannSolver/lbs_01d_initparrays.cc:65]: If num_grps is size_t (which would be consistent since it comes from size()), that should fix it.

**Review by wdhawkins — APPROVED** [review]
